src/order/heyting_algebra.lean

```diff
@@ -0,0 +1,678 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.bounded_order
+
+/-!
+# Heyting algebras
+
+This file defines Heyting algebras.
+
+## Main declarations
+
+* `generalized_heyting_algebra`
+* `generalized_coheyting_algebra`
+* `heyting_algebra`
+* `coheyting_algebra`
+* `biheyting_algebra`
+
+## References
+
+* [Francis Borceux, *Handbook of Categorical Algebra III*][borceux-vol3]
+-/
+
+set_option old_structure_cmd true
+
+open function order_dual
+
+universes u
+variables {ι α β : Type*}
+
+/-! ### Notation -/
+
+/-- Syntax typeclass for Heyting implication. -/
+@[notation_class] class has_himp (α : Type*) := (himp : α → α → α)
+
+/-- Syntax typeclass for Heyting negation. -/
+@[notation_class] class has_hnot (α : Type*) := (hnot : α → α)
+
+export has_himp (himp) has_sdiff (sdiff) has_hnot (hnot)
+
+infixr ` ⇨ `:60 := himp
+prefix `￢`:72 := hnot
```

Comment:
Is this the same notation as negation on `Prop`, by the way? I don't see an obvious alternative so it's probably fine to reuse notation.

Reply:
No, compare `￢` and `¬`. I'm not sure this is the right notation, because Borceux (vol 4) uses `￢`/`¬` (can't tell them apart in print) for the pseudo-complement in a Heyting algebra. On the other hand, he never mentions co-Heyting algebras, so we might be fine?